### PR TITLE
common: Agent bring up on Linux

### PR DESCRIPTION
**Pull request by tomereli**
- Disable son slave watchdog by default which runs system call to restart the son slaves if they crash. We want to catch those, especially with stop_on_failure=1, so this causes confusion when debugging such issues. In addition, we have beerocks_watchdog script which knows if a son slave crashed, so it can take the action.

- Fix console std redirect to create a readable file

- Add stubbed "dummy" bwl

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -25,4 +25,6 @@ tlvf.log
 #klockwork
 kw_reports/*.tmp
 kw_reports/rdkb/*.tmp
-kw_reports/ugw/*.tmp
\ No newline at end of file
+kw_reports/ugw/*.tmp
+
+.vscode/ipch
```
Comment (arnout) [inline .gitignore]: Again, I don't think this is appropriate.

File: beerocks/bcl/source/beerocks_os_utils.cpp [hunk 2]
```diff
@@ -171,7 +171,7 @@ bool os_utils::touch_pid_file(std::string file_path)
 
 int os_utils::redirect_console_std(std::string log_file_name)
 {
-    int fd_log_file_std = open(log_file_name.c_str(), O_CREAT | O_APPEND | O_RDWR, S_IWUSR);
+    int fd_log_file_std = open(log_file_name.c_str(), O_CREAT | O_APPEND | O_RDWR, S_IWUSR | S_IRUSR);
```
Comment (arnout) [inline beerocks/bcl/source/beerocks_os_utils.cpp]: I had noticed this annoying thing as well :-)

Typically, however, I'd create a log file with `S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH`. Also, I usually use `0644` instead of the symbolic names, because I feel that that is much more readable.

File: beerocks/bcl/source/beerocks_os_utils.cpp [hunk 3]
```diff
@@ -171,7 +171,8 @@ bool os_utils::touch_pid_file(std::string file_path)
 
 int os_utils::redirect_console_std(std::string log_file_name)
 {
-    int fd_log_file_std = open(log_file_name.c_str(), O_CREAT | O_APPEND | O_RDWR, S_IWUSR | S_IRUSR);
+    int fd_log_file_std =
+        open(log_file_name.c_str(), O_CREAT | O_APPEND | O_RDWR, S_IWUSR | S_IRUSR);
```
Comment (arnout) [inline beerocks/bcl/source/beerocks_os_utils.cpp]: I prefer whitespace-only changes to be in a separate commit, to make review easier...

File: beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp [hunk 4]
```diff
@@ -6,7 +6,7 @@
  * See LICENSE file for more details.
  */
 
-#include "mon_wlan_hal_dwpal.h"
+#include "mon_wlan_hal_dummy.h"
 
 #include <beerocks/bcl/beerocks_utils.h>
```
Comment (arnout) [inline beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp]: Are all these includes relevant for the dummy implementation?

File: beerocks/bwl/dummy/mon_wlan_hal_dummy.h [hunk 5]
```diff
@@ -52,10 +52,10 @@ class mon_wlan_hal_dwpal : public base_wlan_hal_dwpal, public mon_wlan_hal {
 
     // Private data-members:
 private:
-    std::shared_ptr<char> m_temp_dwpal_value;
+    std::shared_ptr<char> m_temp_dummy_value;
```
Comment (arnout) [inline beerocks/bwl/dummy/mon_wlan_hal_dummy.h]: I don't think this is used anywhere.

**Comment by blogic** [comment]
All commits are missing the SoB line

File: beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp (hunk 4 above)
Comment (tomereli) [reply beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp]: They might not be, but then again I expect the dummy implementation to be deleted once we bring up DWPAL + hwsim on Linux, no point in having 2 stub implementations.

File: beerocks/bwl/dummy/mon_wlan_hal_dummy.h (hunk 5 above)
Comment (tomereli) [reply beerocks/bwl/dummy/mon_wlan_hal_dummy.h]: It is just copy-paste from dwpal bwl without much cleanups other than making everything as stub and having this compile and get to ~operational with controller + agent.
I don't think cleanups are reqyured in the dummy files since we plan to delete this after bringing up dwpal + hwsim.

File: beerocks/bcl/source/beerocks_os_utils.cpp (hunk 3 above)
Comment (tomereli) [reply beerocks/bcl/source/beerocks_os_utils.cpp]: Basically, for easier code review I agree - our normal process is to add commits on top of each other but eventually squash them. I will make sure to follow this rule in future commits (we should have some document that describes the code review process).

File: .gitignore (hunk 1 above)
Comment (tomereli) [reply .gitignore]: What do you mean?

File: beerocks/bcl/source/beerocks_os_utils.cpp (hunk 2 above)
Comment (tomereli) [reply beerocks/bcl/source/beerocks_os_utils.cpp]: Done

File: beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp (hunk 4 above)
Comment (arnout) [reply beerocks/bwl/dummy/mon_wlan_hal_dummy.cpp]: Might make sense to add a comment or readme that this will be removed. Then again, it will be removed so quickly that it's not worth the bother :-)

File: beerocks/bwl/dummy/mon_wlan_hal_dummy.h (hunk 5 above)
Comment (arnout) [reply beerocks/bwl/dummy/mon_wlan_hal_dummy.h]: OK, so I'll lighten up my review.

File: beerocks/bcl/source/network/network_utils.cpp [hunk 6]
```diff
@@ -254,8 +254,8 @@ int network_utils::get_iface_info(network_utils::iface_info &info, const std::st
         }
     }
 
-    if (info.mac.empty()) {
-        network_utils::linux_iface_get_mac(iface_name, info.mac);
+    if (info.mac.empty() && !network_utils::linux_iface_get_mac(iface_name, info.mac)) {
```
Comment (arnout) [inline beerocks/bcl/source/network/network_utils.cpp]: I personally don't like to depend on the short-circuiting of `&&` like this. It's a bit like `goto` IMO. It's acceptable to use it, but only for specific patterns. In this particular case, I think it's confusing because the condition of empty has no direct bearing on returning -1; the return -1 only applies to the `linux_iface_get_mac` call.

Ideally, I like to have a coding rules library with the "approved" patterns for the project.

An example of a short-circuit pattern that I would approve is the following.
```c
if (!foo() ||
    !bar() ||
    memcmp(x, y, sizeof(x))) {
        return -1;
}
```
i.e. a sequence of tests or operations with exit in case of error.

Actually, in C++ I'd even prefer to use exceptions instead of status code returns (in most cases). But since that is not at all the coding style used at the moment, let's stick to status code returns.

Bottom line: in this specific case, I'd write it like this:
```suggestion
    if (info.mac.empty()) {
        if(!network_utils::linux_iface_get_mac(iface_name, info.mac)) {
            return -1;
        }
    }
```

Comment (tomereli) [reply beerocks/bcl/source/network/network_utils.cpp]: Ok, changed. But using exceptions has its drawbacks and it was decided by the original beerocks architect not to use them.
